Add :allow_nil option to delegate_missing_to

CodingAnarchy opened the pull request:
### Summary

Modify `delegate_missing_to` to have an `allow_nil` option and use this for ActiveStorage attachments to address #33016.

CodingAnarchy committed:
Add :allow_nil option to delegate_missing_to; use in ActiveStorage
attachment

georgeclaghorn commented on activestorage/app/models/active_storage/attachment.rb:
```diff
@@ -12,7 +12,7 @@ class ActiveStorage::Attachment < ActiveRecord::Base
   belongs_to :blob, class_name: "ActiveStorage::Blob"
 
-  delegate_missing_to :blob
+  delegate_missing_to :blob, allow_nil: true
```
We consider it a bug for an attachment not to have a blob (see #33405), so I think we can leave this one alone.

CodingAnarchy replied:
Sounds good, I will remove it on that one then. :+1:

georgeclaghorn reviewed (commented):
Thanks for the PR!

georgeclaghorn commented:
There’s a failing test that needs attention:

```
Failure:
ModuleTest#test_delegate_missing_to_raises_delegation_error_if_target_nil [/home/travis/build/rails/rails/activesupport/test/core_ext/module_test.rb:388]:
[Module::DelegationError] exception expected, not
Class: <NameError>
Message: <"undefined local variable or method `target' for #<struct DecoratedTester client=nil>">
---Backtrace---
/home/travis/build/rails/rails/activesupport/test/core_ext/module_test.rb:80:in `method_missing'
/home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/module/delegation.rb:297:in `method_missing'
/home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/module/delegation.rb:300:in `rescue in method_missing'
/home/travis/build/rails/rails/activesupport/lib/active_support/core_ext/module/delegation.rb:296:in `method_missing'
/home/travis/build/rails/rails/activesupport/test/core_ext/module_test.rb:389:in `block in test_delegate_missing_to_raises_delegation_error_if_target_nil'
---------------
```

Schwad commented:
Thanks for doing this @CodingAnarchy ! :)

oniofchaos commented:
👍

lukeredpath commented:
Is there any plan to merge this fix in?

fabriciofreitag commented:
+1

georgeclaghorn commented:
@CodingAnarchy, can you please rebase against latest master and squash your commits? I think we’ll also want a changelog entry for Active Support.

CodingAnarchy commented:
@georgeclaghorn I have now rebased against latest master and squashed the commits to one, as well as included changelog entries for both ActiveSupport and ActiveStorage.  (I can remove the ActiveStorage one if it is viewed as unnecessary.) . Thanks!

georgeclaghorn commented on activestorage/CHANGELOG.md:
````diff
@@ -32,4 +32,18 @@
+
+    delegate :filename, to: :file, allow_nil: true
+  end
+  ```
````
A clearer version:

> Methods calls on singular attachments return `nil` when no file is attached.
>
> Previously, assuming the following `User` model, `user.avatar.filename` would raise a `Module::DelegationError` if no avatar was attached:
>
> ```ruby
> class User < ApplicationRecord
>   has_one_attached :avatar
> end
> ```
>
> They now return `nil`.

Also, new changelog entries get added to the top of the file. Can you move yours up?

georgeclaghorn commented on activestorage/lib/active_storage/attached/many.rb:
```diff
@@ -3,7 +3,7 @@
   # Decorated proxy object representing of multiple attachments to a model.
   class Attached::Many < Attached
-    delegate_missing_to :attachments
+    delegate_missing_to :attachments, allow_nil: true
```
`:allow_nil` isn’t necessary in this case. `attachments` will always return an `Array` or `ActiveRecord::Associations::CollectionProxy`.

Nsouza31 replied:
Hi George I hope you are fine!

I am building an App where I have two kinds of users (its a platform) and in one of this cases the Avatar must be compulsory in other hand for the end-user it is optional. Given the fact that I am tried to minimize the numbers of fields in the form to avoid registration abandon.

**Important**: I am use has_one_attached 
After read the discussion above I assume that  the behavior "  delegate_missing_to :attachment, allow_nil: true" is already in place when I ran the command "rails active_storage:install". Is that true?

My actual behavior:
When I try to create a user without avatar, it is created, but the active recorde returns an error:

  User Create (0.5ms)  INSERT INTO "users" ("encrypted_password", "created_at", "updated_at", "cpf", "full_name", "phone_number", "is_company", "status", "country", "authentication_token") VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10) RETURNING "id"  [["encrypted_password", "$2a$11$dmNXTW1etZKXqpLhvhT0c.7FFv5V4u3TffsyLgInIHKFtMO4zd5K6"], ["created_at", "2020-02-10 23:13:06.886745"], ["updated_at", "2020-02-10 23:13:06.886745"], ["cpf", "62056594099"], ["full_name", "Silvia Brandao"], ["phone_number", "51998909909"], ["is_company", false], ["status", "Ativo"], ["country", "BR"], ["authentication_token", "v-HXKQxPrTPSxoGBoApX"]]
  ↳ app/controllers/users/registrations_controller.rb:15:in `create'
   (1.1ms)  COMMIT
  ↳ app/controllers/users/registrations_controller.rb:15:in `create'
  User Load (0.9ms)  SELECT "users".* FROM "users" WHERE "users"."id" = $1 ORDER BY "users"."id" ASC LIMIT $2  [["id", 8], ["LIMIT", 1]]
  ↳ app/controllers/users/registrations_controller.rb:15:in `create'
  ActiveStorage::Attachment Load (1.7ms)  SELECT "active_storage_attachments".* FROM "active_storage_attachments" WHERE "active_storage_attachments"."record_id" = $1 AND "active_storage_attachments"."record_type" = $2 AND "active_storage_attachments"."name" = $3 LIMIT $4  [["record_id", 9], ["record_type", "User"], ["name", "photo"], ["LIMIT", 1]]
  ↳ app/controllers/users/registrations_controller.rb:15:in `create'
Completed 500 Internal Server Error in 332ms (ActiveRecord: 12.1ms | Allocations: 33457)

Db:shema:
create_table "users", force: :cascade do |t|
    t.string "email", default: ""
    t.string "encrypted_password", default: "", null: false
    t.string "reset_password_token"
    t.datetime "reset_password_sent_at"
    t.datetime "remember_created_at"
    t.datetime "created_at", precision: 6, null: false
    t.datetime "updated_at", precision: 6, null: false
    t.string "cpf"
    t.string "cnpj"
    t.string "full_name"    
    t.integer "photo", array: true
end

I would realy appreciate your help!

georgeclaghorn commented on activesupport/CHANGELOG.md:
```diff
@@ -19,5 +19,10 @@
+    Now the `allow_nil` option has been added to enable the user to specify they want `nil` returned
+    in this case.
+
+    *Matthew Tanous*
```
Please move this to the top of the file. Wrap changelog lines to 80 characters.

georgeclaghorn reviewed (commented):
Last few things.

CodingAnarchy commented:
@georgeclaghorn I've moved the changelog entries to the top of the files, and removed the `allow_nil` option from `activestorage/lib/active_storage/attached/many.rb` as requested.

georgeclaghorn commented:
Thank you!
